### DISCUSSION TITLE
Custom Node and updateNodeInternals not firing to change handle position

### DISCUSSION THREAD

**Discussion by FA5I** [xyflow/xyflow#4307] [Q&A] [answered]
Hi,

I have a custom node and an Elkjs layout. Upon switching layouts, I want the handle position to change from left/right to top/bottom.

However, when in the following code example, the position is not changing. Can someone please assist and explain what I could do to get it to work?:

Upon clicking "vertical layout" change the position of the handle.

[example code](https://codesandbox.io/p/sandbox/focused-galileo-mk9747?file=%2FApp.js%3A68%2C45&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clwpbj68000072e6a4h5zwsq7%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clwpbj68000032e6ac2uwa1vg%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clwpbj68000042e6a4c1qsqbc%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clwpbj68000062e6aqovln66q%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B62.237664601306676%252C37.762335398693324%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clwpbj68000032e6ac2uwa1vg%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clwpbj68000022e6ax4ha8mhf%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FApp.js%2522%252C%2522state%2522%253A%2522IDLE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A68%252C%2522startColumn%2522%253A45%252C%2522endLineNumber%2522%253A68%252C%2522endColumn%2522%253A45%257D%255D%257D%252C%257B%2522id%2522%253A%2522clwpc56ig00022e6a7x2j8smz%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A20%252C%2522startColumn%2522%253A24%252C%2522endLineNumber%2522%253A20%252C%2522endColumn%2522%253A24%257D%255D%252C%2522filepath%2522%253A%2522%252FTextUpdaterNode.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clwpbj68000032e6ac2uwa1vg%2522%252C%2522activeTabId%2522%253A%2522clwpbj68000022e6ax4ha8mhf%2522%257D%252C%2522clwpbj68000062e6aqovln66q%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clwpbj68000052e6afa6nmpon%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clwpbj68000062e6aqovln66q%2522%252C%2522activeTabId%2522%253A%2522clwpbj68000052e6afa6nmpon%2522%257D%252C%2522clwpbj68000042e6a4c1qsqbc%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clwpbj68000042e6a4c1qsqbc%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

**Comment by bcakmakoglu** [comment]
You have a couple of issues in your sandbox.

First of all, the most important issue here, you're putting properties onto the node object that don't belong there.
```ts
{
    id: "1",
    type: "textUpdater",
    position: { x: 0, y: 0 },
    vertical_positon: false,  // <---- this is a no-no, put it into the `data` object below!
    data: { value: 123 },
  },
```

Any data that you need for your nodes belongs in the `node.data` object, otherwise there is no guarantee that ReactFlow doesn't prune them away internally.

```ts
{
    id: "1",
    type: "textUpdater",
    position: { x: 0, y: 0 },
    data: { value: 123, vertical_positon: false }, // now you're safe!
  },
```

Additionally this property that you added to the node **will not be passed to your custom node**.
Since it's not a property known to ReactFlow, it won't be automatically forwarded to your custom node, so you can't do `props.vertical_position` in your custom nodes `useEffect`.

```ts
  useEffect(() => {
    // ... your code
  }, [props.vertical_mode]); // <--- this is not a prop you will ever receive!
```

That is what the `data` object is for - after moving your property into the data object, you can access it from there

```ts
  useEffect(() => {
    // ... your code
  }, [props.data.vertical_mode]); // <--- now you can access the property
```

The next issue is that you're not using the same property name in the node definition and custom node.
Notice that the property on the node was called `vertical_position` whereas the property in the `useEffect` is called `vertical_mode`.
These names need to match up of course - so use one name, maybe something simple like `vertical`.

[Here's a corrected sandbox.](https://codesandbox.io/p/sandbox/nostalgic-cherry-ynkvdz?file=%2FTextUpdaterNode.js%3A10%2C1-20%2C29)

Although I'd maybe recommend using a different approach to pass the vertical mode down to your nodes - a context possibly instead of an extra property in the `node.data` object that you'd have to constantly modify for all nodes.

Something like this (just an example of course)
```tsx
<LayoutModeProvider>
	<ReactFlow ...>
		// ... rest
```

```tsx
// CustomNode.tsx

export default function CustomNode() {
	const isVertical = useIsVertical();
}

**Reply by FA5I** [reply]
Thank you @bcakmakoglu ,

That does update the  layout but there is a slight quirk. If I pan the canvas a little bit before hitting the layout update button, the layout handle position does not change. In my actual app, this shows as an edge connecting to the side. Any ideas as to what it could be? Seems like there may be some need to sync the updates



https://github.com/xyflow/xyflow/assets/56031768/94dbaa2a-d72a-4241-9a5f-1b9a3aea3086

**Reply by bcakmakoglu** [reply, marked as answer]
Works fine for me 🤔

https://github.com/xyflow/xyflow/assets/78412429/dec57fbb-221d-457e-aabc-a5bfe4ade6a6

**Reply by FA5I** [reply]
Thank you!

Hmmn that's interesting. Wonder if it's to do with the browser. I'll play around with a couple and see if impacts it. For now I used this technique and it works without the lag:

https://reactflow.dev/examples/layout/elkjs
